### PR TITLE
Rewrite and improvements (with new command)

### LINKED ISSUE #29: [FEATURE] Add a help command
A help command for the bot will be useful.

### LINKED ISSUE #36: [FEATURE] Create dynamic help commands
**Is your feature request related to a problem? Please describe.**
In BotANF v1, the help command was almost fully hard coded which is not very good to maintain and code-quality was very low.

**Describe the solution you'd like**
The help command be as dynamic as it could.

**Additional context**
- The command sorting should be handled
- The commands per embed displayed should be handled
- The command categories should be handled
- Hidden commands should not be displayed
- Descriptions should be shown alongside
- A special help command i.e `help <command name>` should be set up the same way.

### PR DESCRIPTION
**Pull request by Oseryx**
# Rewrite and Improvements

## Description

- Addition of discord buttons
- Addition of Arguments
- Addition of a Help Command
- Rewrite of the client class, the command handler, the event handler and the logger class

These changes allows a better code maintainability and provide better options for commands 

Fixes #36 
Fixes #29

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Details

I know this section isn't the template, but I need to explain the changes

### Arguments:

A new class that parses the message with the given arguments, it takes and array of Argument Options type, an Argument Options can take up to 5 options:
- id: name of argument, it's also the variable name you'll receive after the parsing
- type: can be a number, string, member, user, channel or a role
- description (optional): a description of the argument, useful for the help command if details are needed to be provided
- nullable (optional): if it's true then the part of message that should be parsed will be left as it is
- default (optional): default value for nullable options if needed

Suppose you have two args `channel` and `member` the class will return: 
`{ channel: GuildChannel, member; GuildMember, args: string[] }`

### Commands:

Rewrite of the command base class, it get much more options (They're all optional) :
- aliases
-  description (optional): useful for the help command
- usage: usage example of the command, purely made for the help command
- args: already explained above
- ownerOnly: commands only for set of users (specified in the options of the Client class)
- category: category of the command, if not specified it'll set in the `default` category
- clientPermissions: permissions required by the bot to execute the command
- userPermissions: permissions required for the user to execute the command
- channel: where the command should live like `dm` or `guild text channel` (not implemented yet)

#### Command Handler:
It handles the options of commands args and loads everything, missing features (it only loads an entire folder) :
- Unload command
- Load command

### Events and event handler:

Nothing new the same event handler but rewritten, missing features (it only loads an entire folder) :
- Unload event
- Load event

### Client:

Rewrite of the client class, it has 4 properties:
- options: it has only the ownerId which is an array of strings
- command: Collection of the commands
- events: collection of the events
- categoies: Collection of the categories with their commands (purely made for the help command)

Methods: isOwner check if a user is one of the bot owners

### Logger:

Rewrite of the entire logger class, no instance is needed to be created everything is static as logger should be, missing feature:
- Show where the log comes from (it'll be added soon passing the filename as an argument is just ugly)

Note:  A lot of the old code has been commented, to make everything works without mongodb (another pr will be made for it)

### Images

Images for the command help only:

image
image
image

For the rewrite what is really changing is the code readability before you specify `super(...)` without knowing where you are putting your values exactly now it's way more clearer:

image

## How Has This Been Tested?

I've tested the handlers (creating folders inside folders for example) it handles module loading, I've also tested argument parsing with nullable arguments and yet everything is working perfectly

Since this project doesn't have unit testing library in the dependencies I decided to don't add one, the choice is up to you

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -14,22 +14,22 @@ orientation.
...
-- Other conduct which could reasonably be considered inappropriate in a
-  professional setting
+-   The use of sexualized language or imagery and unwelcome sexual attention or
+    advances
+-   Trolling, insulting/derogatory comments, and personal or political attacks
+-   Public or private harassment
+-   Publishing others' private information, such as a physical or electronic
+    address, without explicit permission
+-   Other conduct which could reasonably be considered inappropriate in a
+    professional setting
```
Comment (anf) [inline CODE_OF_CONDUCT.md]: Why are there extra spaces here?

File: PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -16,10 +16,10 @@ Fixes # (issue number)
...
 -->
 
-- [ ] Bug fix (non-breaking change which fixes an issue)
-- [ ] New feature (non-breaking change which adds functionality)
-- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-- [ ] This change requires a documentation update
+-   [ ] Bug fix (non-breaking change which fixes an issue)
+-   [ ] New feature (non-breaking change which adds functionality)
+-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
+-   [ ] This change requires a documentation update
```
Comment (anf) [inline PULL_REQUEST_TEMPLATE.md]: Why are there extra spaces here?

File: PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -33,20 +33,20 @@ Images to show to preview of the change, how it looks and effects the code.
 Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.
 -->
 
-- [ ] Test A
-- [ ] Test B
+-   [ ] Test A
+-   [ ] Test B
```
Comment (anf) [inline PULL_REQUEST_TEMPLATE.md]: Why are there extra spaces here?

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (Oseryx) [reply CODE_OF_CONDUCT.md]: I didn't touch the file, maybe it's prettier who added an extra space

File: PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (Oseryx) [reply PULL_REQUEST_TEMPLATE.md]: I didn't touch the file, prettier certainly did

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (anf) [reply CODE_OF_CONDUCT.md]: Please run `npm run format` and that should revert these changes, if not. Please do something about it and revert all these extra spaces.

Comment (Oseryx) [reply CODE_OF_CONDUCT.md]: Sure

File: src/utils/commands/CommandHandler.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+                command.exec(message, parsedArguments.parsedArguments);
+            } else {
+                //TODO: implement embed to send whenever the user isn't allowed to run the command
+                //message.channel.send(`You're not the owner`);
+            }
+        }
+    }
+
+    private checkForPermissions(message: Message, command: Command): boolean {
```
Comment (anf) [inline src/utils/commands/CommandHandler.ts:66]: I think a better name for this function would be; "isAuthorized" or "userIsAuthorized".

Comment (Oseryx) [reply src/utils/commands/CommandHandler.ts:66]: or maybe `checkForUserPermissions`

Comment (anf) [reply src/utils/commands/CommandHandler.ts:66]: I think that shorter version might get along better.

**Pull request closed without merging**